Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs

```diff
@@ -0,0 +1,211 @@
+        {
+            _productSearchService = productSearchService;
+            _categorySearchService = categorySearchService;
+            _propertySearchService = propertySearchService;
+            _propertyDictionarySearchService = propertyDictionarySearchService;
+            _catalogSearchService = catalogSearchService;
+        }
+
+        public int TotalCount { get; set; }
+        public int CurrentPageNumber { get; protected set; }
+        public int PageSize { get; set; } = 50;
+
+        public IEnumerable<IExportable> Items { get; protected set; }
+
+        private bool _statesInitialized = false;
+
+        private readonly ExportDataSourceState[] _exportDataSourceStates = new ExportDataSourceState[5];
+
+        private ExportDataQuery _dataQuery;
+
+        public ExportDataQuery DataQuery
+        {
+            set
+            {
+                _dataQuery = value;
+                CurrentPageNumber = 0;
+                TotalCount = -1;
+                InitExportDataSourceStates();
+                CalculateCounts();
+            }
+        }
+
+        public int GetTotalCount()
+        {
+            CalculateCounts();
+            return TotalCount;
+        }
+
+        private void EnsureHaveTotals()
+        {
+            if (TotalCount < 0)
+            {
+                CalculateCounts();
+            }
+        }
+
+        public bool Fetch()
+        {
+            EnsureHaveTotals();
+
+            var take = _dataQuery.Take ?? PageSize;
+            var skip = _dataQuery.Skip ?? CurrentPageNumber * PageSize;
+            var taskList = new List<Task>();
+            Items = new List<IExportable>();
+
+            var curStateStartPos = 0; // Starting position of current state
+
+            foreach (var state in _exportDataSourceStates)
+            {
+                state.FetchResult.Result = Array.Empty<IExportable>();
+                var overlap = skip - curStateStartPos; // Positive overlap should fall into skip, negative into take
+                state.SearchCriteria.Skip = overlap < 0 ? 0 : overlap;
+                state.SearchCriteria.Take = (skip + take > curStateStartPos + state.FetchResult.TotalCount) ? curStateStartPos + state.FetchResult.TotalCount - skip : take + (overlap < 0 ? overlap : 0);
+
+                if (state.SearchCriteria.Take > 0) // Skip running this fetch in case of nothing to take
+                {
+                    taskList.Add(state.FetchFunc(state));
+                }
+
+                curStateStartPos += state.FetchResult.TotalCount;
+            }
+
+            Task.WhenAll(taskList).GetAwaiter().GetResult(); // Ensures completing all fetches
+
+            ((List<IExportable>)Items).AddRange(_exportDataSourceStates.SelectMany(x => x.FetchResult.Result));
+            CurrentPageNumber++;
+
+            return Items.Any();
+        }
+
+        private void CalculateCounts()
```

Comment:
The idea of this method is to have all states TotalCount filled, not just to get overall TotalCount:
https://github.com/VirtoCommerce/vc-platform-core/pull/521/files#diff-025b92c5e3684f6c6184640d95e9ba6dR120.
Thus we can use these counts in future calculation which will simplify paging logic

Reply:
There is the same, because each of FetchFunc fill TotalCount for state

Reply:
Really, I see this now.